README.md
# HomeApp

## Overview

HomeApp is an application for handling I/O for Home Database which are designed to contains personal and private
information. Motivations for creating this project can be read on my personal
[blog post](https://dskrzypiec.dev/home-db/).

If you want to check HomeApp without compiling the project locally, you can visit

* [https://homeappdev.dskrzypiec.dev](https://homeappdev.dskrzypiec.dev) or
    [http://3.74.170.81](http://3.74.170.81)
* and use credentials `testuser/password` to log in

## Build

On Linux or MacOS simply do

```
bash build.sh
```

On other platforms you can build directly using Go compiler:

```
go generate
go build
```

Running the HomeApp by `./homeApp`. As default will not include 2FA via Telegram and will use test SQLite database
(`test.db`).

### Credentials

For test database (`test.db`) there is single user:

```
login: testuser
password: password
```

### Flags

HomeApp supports the following options

* `-dbPath path` - path to Home Database
* `-port 8080` - port on which HomeApp will be listening
* `-telegram2fa` - if enabled, then HomeApp will use two-factor authentication (2FA) using Telegram channel. More details below.
* `-publishViewsAfter 300` - numbers of minutes after which endpoints view
      statistics will be published. If Telegram is configured, then it'll be sent
      over the Telegram channel. Otherwise just logged
* `-logDebug` - flag for enabling debug log level. If disabled, then "info" log level would be used
* `-logConsole` - flag for using `ConsoleWriter` within `zerolog`. Convenient for local development but is less efficient then standard writer.


### 2FA via Telegram

In case when 2FA via Telegram is enabled you have to provide the following environment variables:

```
export HOMEAPP_TELEGRAM_CHANNEL_ID=-100xxxxxxxx65
export HOMEAPP_TELEGRAM_BOT_TOKEN=11111111:AAAAAA-jjjjj-vTDSFSKDFndfkldsG
```

Values of those variables came from setting up [Telegram bot](https://core.telegram.org/bots/api) and dedicated Telegram
channel for the communication.


## High level design

**TODO**


CHANGELOG.md
# 0.5.3
    * Go back to 100 MB document file limit. Current EC2 is too small for that.

# 0.5.2
    * Increased document file size from 100 MB to 275 MB

# 0.5.1
    * Changed column chart to bar chart in Financial Explorer

# 0.5.0
    * First working version with Finance Explorer - monthly aggregated chart
      for filtered transactions.

# v0.4.2
    * Fix hardcoded 2022 year in financial summary

# v0.4.1
    * Add flags for logger

# v0.4.0
    * Add e-books and paper books info summary on the /home page

# v0.3.0
    * Add user session timer on /home page
    * Add functionality of user session prolongation by click onto user session
      timer on /home page

# v0.2.2
    * Extend fields (in the form and database) for books - category and language
    * Add sending Telegram message after new book entry was inserted

# v0.2.1
    * Make dynamic number of columns in table of documents based on screen size
    * Make dynamic number of columns in table of books based on screen size

# v0.2.0
    * Add (e)books page
    * Books (just info about a book) or e-book (info + e-book file) can be
      added
    * Books can be browsed and filtered
    * Books, in case of e-books, can be downloaded

# v0.1.4
    * Update monitoring page view counts by putting unregistered endpoint paths
      into single category, to make published statistics more readable (there
      are a lots of bots and crawlers out there!)
    * JWT signing key is now randomly generated

# v0.1.3
    * Add basic monitoring regarding page view counts. If Telegram 2FA is
      enabled monitoring messages would be send onto Telegram channel.

# v0.1.2
    * If 2FA is enabled, Telegram messages will be sent after insertions into
      database (counters state, financial transactions and documents)

# v0.1.1
    * Extend configuration (port)
    * Embed application version and git commit SHA into footer on /home
    * Add ./build.sh to build application (`go generate` is not needed before `go build`)
    * Initialized README

# v0.1.0
    * First deployed production version
    * Home database summary on /home page
    * Financial summary
    * Financial uploader for PKO Bank XML transactions history files
    * Documents listing
    * Uploading new documents
    * Possible 2FA via Telegram channel


config.go
package main

import (
	_ "embed"
	"flag"
	"os"
	"strings"
	"time"

	"github.com/rs/zerolog"
	"github.com/rs/zerolog/log"
)

const (
	SessionTimeoutMinutes = 10
	TelegramBotTokenEnv   = "HOMEAPP_TELEGRAM_BOT_TOKEN"
	TelegramChannelIdEnv  = "HOMEAPP_TELEGRAM_CHANNEL_ID"
)

//go:generate sh -c "head -1 CHANGELOG.md > VERSION.txt"
//go:generate sh -c "printf %s $(git rev-parse HEAD) >> VERSION.txt"
//go:embed VERSION.txt
var versionFile string

type Config struct {
	Port                  int
	DatabasePath          string
	UseTelegram2FA        bool
	Telegram              *TelegramConfig
	SessionTimeoutMinutes int
	HttpClientTimeout     time.Duration
	Logger                LoggerConfig
	AppVersion            string
	CurrentCommitSHA      string
	PublishViewsAfter     time.Duration
}

type TelegramConfig struct {
	BotToken  string
	ChannelId string
}

type LoggerConfig struct {
	UseDebugLevel    bool
	UseConsoleWriter bool
}

// Parse or fail.
func ParseConfigFlags() Config {
	telegram2fa := flag.Bool("telegram2fa", false, "Use Telegram for two-factor authentication")
	dbPath := flag.String("dbPath", "test.db", "Path to SQLite Home DB")
	port := flag.Int("port", 8080, "Port on which HomeApp is listening")
	publishViewsAfter := flag.Int("publishViewsAfter", 300,
		"After each 'x' minutes endpoints views statistics will be published")

	logDebugLevel := flag.Bool("logDebug", true,
		"Log events on at least debug level. Otherwise info level is assumed.")
	logUseConsoleWriter := flag.Bool("logConsole", true,
		"Use ConsoleWriter within zerolog - pretty but not efficient, mostly for development")
	flag.Parse()

	var telegramConfig *TelegramConfig

	if *telegram2fa {
		telegramBotToken := os.Getenv(TelegramBotTokenEnv)
		telegramChannelId := os.Getenv(TelegramChannelIdEnv)
		if telegramBotToken == "" {
			log.Fatal().Msgf("[config] Telegram 2FA is on, %s env variable should be set", TelegramBotTokenEnv)
		}
		if telegramChannelId == "" {
			log.Fatal().Msgf("[config] Telegram 2FA is on, %s env variable should be set", TelegramChannelIdEnv)
		}
		telegramConfig = &TelegramConfig{
			BotToken:  telegramBotToken,
			ChannelId: telegramChannelId,
		}
	}

	loggerConfig := LoggerConfig{
		UseDebugLevel:    *logDebugLevel,
		UseConsoleWriter: *logUseConsoleWriter,
	}

	appVersion, commitSha := parseVersions(versionFile)

	return Config{
		Port:           *port,
		DatabasePath:   *dbPath,
		UseTelegram2FA: *telegram2fa,
		Telegram:       telegramConfig,

		SessionTimeoutMinutes: SessionTimeoutMinutes,
		HttpClientTimeout:     60 * time.Second,
		Logger:                loggerConfig,

		PublishViewsAfter: time.Duration(*publishViewsAfter) * time.Minute,

		AppVersion:       appVersion,
		CurrentCommitSHA: commitSha,
	}
}

func (c *Config) setupZerolog() {
	zerolog.DurationFieldUnit = time.Millisecond
	if c.Logger.UseDebugLevel {
		zerolog.SetGlobalLevel(zerolog.DebugLevel)
	}
	if c.Logger.UseConsoleWriter {
		log.Logger = log.Output(zerolog.ConsoleWriter{Out: os.Stderr, TimeFormat: time.RFC3339})
	} else {
		log.Logger = zerolog.New(os.Stdout).With().Timestamp().Logger()
		zerolog.TimeFieldFormat = time.RFC3339
	}
}

func parseVersions(input string) (string, string) {
	const firstSHAChars = 8

	lines := strings.Split(input, "\n")
	if len(lines) != 2 {
		log.Fatal().Msg("[config] VERSION.txt file should contain exactly 2 lines - app version and commit SHA")
	}
	_, appVersion, appVerOk := strings.Cut(lines[0], " ")
	if !appVerOk {
		log.Fatal().Msgf("[config] incorrect app version. Should be taken from CHANGELOG in form of '# v0.1.0', got: %s",
			lines[0])
	}
	commitSha := lines[1]
	if len(commitSha) <= firstSHAChars {
		log.Fatal().Msgf("[config] incorrect commit SHA length, got: %d", len(commitSha))
	}
	shortSha := lines[1][:firstSHAChars] + "..."

	return appVersion, shortSha
}
